### PR TITLE
feat: Add inverse inverse root functions

### PR DESCRIPTION
**Pull request by bensuperpc**
#### Description of Change
Add inverse inverse root functions (Template version)

- **Fast_InvSqrt** : C++ version of fast inverse square root implementation.
- **Standard_InvSqrt** : The fast inverse square root with standard lib (cmath).

All function are template function and C++11.

#### Checklist
- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1570"></a>

### CODE REVIEW

File: math/inv_sqrt.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        y = y * (1.5 - (x2 * y * y));
+    }
+    return y;
+}
+
+/**
+ * @brief This is the function that calculates the fast inverse square root.
+ * The following code is the fast inverse square root with standard lib (cmath)
+ * More info:
+ * https://www.linkedin.com/pulse/fast-inverse-square-root-still-armin-kassemi-langroodi
```
Comment (Panquesito7) [inline math/inv_sqrt.cpp:47]:
```suggestion
 * More information can be found at
 * [LinkedIn](https://www.linkedin.com/pulse/fast-inverse-square-root-still-armin-kassemi-langroodi)
```

File: math/inv_sqrt.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
+/**
+ * @file
+ * @brief Calculate the inverse inverse root.
```
Comment (Panquesito7) [inline math/inv_sqrt.cpp:3]:
```suggestion
 * @brief Implementation of [Fast Inverse Square Root](https://medium.com/hard-mode/the-legendary-fast-inverse-square-root-e51fee3b49d9).
```

File: math/inv_sqrt.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+ * @details
+ * Two implementation to calculate inverse inverse root,
+ * from Quake III Arena (C++ version) and with a standard library (`cmath`)
```
Comment (Panquesito7) [inline math/inv_sqrt.cpp:6]: Maybe provide more details: how the algorithm works, what does it do, etc..

File: math/inv_sqrt.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ * @tparam T floating type
+ * @param x value to calculate
+ * @return T return inverse square root
+ */
+template <typename T>
+T Standard_InvSqrt(T number) {
+    T squareRoot = sqrt(number);
+    return 1.0f / squareRoot;
+}
+
```
Comment (Panquesito7) [inline math/inv_sqrt.cpp:57]: Please add self-test implementations using `assert` (see the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program as an example).

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: math/inv_sqrt.cpp [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+ */
+
+#include <cmath>        /// for `std::sqrt`
+#include <iostream>     /// for IO operations
+#include <type_traits>  /// for `static_assert`
+
+/**
+ * @brief This is the function that calculates the fast inverse square root.
+ * The following code is the fast inverse square root implementation from
+ * Quake III Arena (Adapted for C++) More information can be found at
```
Comment (Panquesito7) [inline math/inv_sqrt.cpp:17]:
```suggestion
 * Quake III Arena (Adapted for C++). More information can be found at
```

File: math/inv_sqrt.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+ * [Wikipedia](https://en.wikipedia.org/wiki/Fast_inverse_square_root)
+ * @tparam T floating type
+ * @tparam iterations inverse square root, the greater the number of
+ * iterations, the more exact the result will be (1 or 2).
+ * @param x value to calculate
+ * @return T return inverse square root
```
Comment (Panquesito7) [inline math/inv_sqrt.cpp:23]:
```suggestion
 * @return the inverse square root
```

File: math/inv_sqrt.cpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    return y;
+}
+
+/**
+ * @brief This is the function that calculates the fast inverse square root.
+ * The following code is the fast inverse square root with standard lib (cmath)
+ * More information can be found at
+ * [LinkedIn](https://www.linkedin.com/pulse/fast-inverse-square-root-still-armin-kassemi-langroodi)
+ * @tparam T floating type
+ * @param x value to calculate
```
Comment (Panquesito7) [inline math/inv_sqrt.cpp:50]:
```suggestion
 * @param number value to calculate
```

File: math/inv_sqrt.cpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+ * @return T return inverse square root
```
Comment (Panquesito7) [inline math/inv_sqrt.cpp:51]:
```suggestion
 * @return the inverse square root
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: math/inv_sqrt.cpp (hunk 4 above)
Comment (bensuperpc) [reply math/inv_sqrt.cpp:60]: Tests have been added

File: math/inv_sqrt.cpp [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    i = (sizeof(T) == 8 ? 0x5fe6eb50c7b537a9 : 0x5f3759df) - (i >> 1);
+    y = *(T *)&i;
+
+    y = *reinterpret_cast<T *>(&i);
+
+    y = y * (1.5 - (x2 * y * y));
+    if (iterations == 2) {
+        y = y * (1.5 - (x2 * y * y));
+    }
+    return y;
```
Comment (Panquesito7) [inline math/inv_sqrt.cpp:43]: Please add documentation of what this code does here.

File: math/inv_sqrt.cpp [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    assert(std::fabs(Fast_InvSqrt(12.0f) - 0.288423) < epsilon);
+    assert(std::fabs(Fast_InvSqrt<double>(5.0f) - 0.447141) < epsilon);
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();
```
Comment (Panquesito7) [inline math/inv_sqrt.cpp:85]:
```suggestion
    test();  // run self-test implementations
```

File: math/inv_sqrt.cpp (hunk 9 above)
Comment (bensuperpc) [reply math/inv_sqrt.cpp]: > Please add documentation of what this code does here.

Comments have been added

File: DIRECTORY.md [hunk 11]
```diff
@@ -330,5 +331,5 @@
   * [Brute Force String Searching](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/strings/brute_force_string_searching.cpp)
   * [Horspool](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/strings/horspool.cpp)
   * [Knuth Morris Pratt](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/strings/knuth_morris_pratt.cpp)
-  * [Manacher's Algorithm](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/strings/manacher_algorithm.cpp)
-  * [Rabin Karp](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/strings/rabin_karp.cpp)
\ No newline at end of file
+  * [Manacher Algorithm](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/strings/manacher_algorithm.cpp)
```
Comment (Panquesito7) [inline DIRECTORY.md:334]: Did you modify this file manually?

**Review by Panquesito7 — APPROVED** [review]
Awesome work, @bensuperpc! It's been great to be your first contribution here! We hope you keep contributing! 😄👍🎉

**Comment by bensuperpc** [comment]
> Awesome work, @bensuperpc! It's been great to be your first contribution here! We hope you keep contributing!

Thank you :)

**Review by ayaankhan98 — APPROVED** [review]
LGTM :+1:

**Pull request merged**